python/pyarrow/tests/test_pandas.py

```diff
@@ -4082,6 +4082,66 @@ def test_array_to_pandas():
         # tm.assert_series_equal(result, expected)
 
 
+def test_to_pandas_types_mapper():
+    # https://issues.apache.org/jira/browse/ARROW-9664
+    if Version(pd.__version__) < Version("1.0.0"):
+        pytest.skip("ExtensionDtype to_pandas method missing")
+
+    data = pa.array([1, 2, 3], pa.int64())
+
+    # Test with mapper function
+    types_mapper = {pa.int64(): pd.Int64Dtype()}.get
+    result = data.to_pandas(types_mapper=types_mapper)
+    assert result.dtype == types_mapper(data.type)
+
+    # Test mapper function returning None
+    types_mapper = {pa.int64(): None}.get
+    result = data.to_pandas(types_mapper=types_mapper)
+    assert result.dtype == data.type.to_pandas_dtype()
+
+    # Test mapper function not containing the dtype
+    types_mapper = {pa.float64(): pd.Float64Dtype()}.get
+    result = data.to_pandas(types_mapper=types_mapper)
+    assert result.dtype == data.type.to_pandas_dtype()
+
+    # Test for the interval extension dtype
+    # -> ignores mapping and uses default conversion
+    types_mapper = {pa.float64(): pd.IntervalDtype()}.get
+    result = data.to_pandas(types_mapper=types_mapper)
```

Comment:
I have been thinking about this before and am not sure what is the best way. I agree that it doesn't really makes that much sense to have `types_mapper` for Array as it does for the Table. `dtype` kind of a think should do.

If I use the example for the intervaldtype you mentioned, it works fine without the need for a `types_mapper`, so I am not sure what should be done here (you can't change to any other dtype anyways?):

```
>>> interval = pd.Series(pd.interval_range(0, 5, 5))
>>> data = pa.array(interval)
>>> data.to_pandas()
0    (0, 1]
1    (1, 2]
2    (2, 3]
3    (3, 4]
4    (4, 5]
dtype: interval
````

I was thinking in general about intervaldtype or any other Pandas extension dtype and the only thing I could think of is to add:
```
if hasattr(types_mapper(dtype), '__from_arrow__'):
            array = types_mapper(dtype).__from_arrow__(array)
```
but that didn't really work as the array is a numpy array at that point (`_array_like_to_pandas()`) - if I remember correctly.

Reply:
> If I use the example for the intervaldtype you mentioned, it works fine without the need for a `types_mapper`, so I am not sure what should be done here (you can't change to any other dtype anyways?):

Ah, yes, I forgot about that this already works. That's because it is also an extension type on the Arrow side. A way to test is would be to do `data.storage.to_pandas(..)`, but then it is the same as the other tests with Int64Dtype, so no need to include it.

Reply:
So from an offline discussion: conclusion is that we can just remove this additional test case with interval dtype, since using the storage type would not work with pandas, and converting a plain pyarrow array to a pandas extension dtype is already covered by the int64 case above.